### DISCUSSION TITLE
Icons on the left?

### DISCUSSION THREAD

**Discussion by muellerto** [leo-arch/clifm#309] [General]
Question: What's this?

![1](https://github.com/user-attachments/assets/8ecbc214-8ece-493b-831a-64fd192b8561)

Why are they all always equal? And why do I suddenly get these icons when I use parameter `-E` ?

The picture is from my Windows machine. On Linux I can use `-E` without getting this effect. What does this depend on?

BTW: I searched for an explanation in the wiki and in the FAQ. Seems that a documentation of the screen in both view modes (short and long), it's elements and the settings the elements depend on is missing.

**Comment by leo-arch** [comment]
Hi @muellerto. The problem is simple: workspace colors are mainly used in the prompt, where we need to use `\001` and `\002` to mark the color string as a series non-printing characters. The thing is that some terminal do not like these codes and print garbage instead (those are indeed no icons). I'll fix it as soon as I can.

**Comment by leo-arch** [comment]
Should be fixed now.

**Comment by muellerto** [comment]
Yes, the none-icons are gone. I had this indeed in every "terminal" I tried on Windows (mintty, alacritty, wt, hyper, ConEmu, DOS-prompt).

OK, meanwhile I have `autocmd lines configured which make different color schemes for different workspaces.

**Comment by muellerto** [comment]
I found some more. I deleted a file named t, then clifm makes such a display:

![1](https://github.com/user-attachments/assets/7d0b0672-21f5-40f5-bfa7-c832976d0bd8)

**Reply by leo-arch** [reply]
Should be fixed now.

**Reply by muellerto** [reply]
Works good! I have only the green checkmark. Think this is intended.

**Reply by leo-arch** [reply]
Yes, it is intended. It's part of the new Unicode interface.

**Comment by leo-arch** [comment]
Btw, we have a new columns arrangement algorithm: still not as good as **ls**(1), but much better than before nonetheless. It's still an experimental, and thereby hidden feature. Compile as follows to enable it:

```sh
export CPPFLAGS="$CPPFLAGS -DTIGHT_COLUMNS"
make
```
Give it a try and let me know if something comes out. I think it's a pretty neat improvement.

**Reply by muellerto** [reply]
In short view mode? Looks not bad. Leads to less rows, so you get more on a screen when you have a lot of files and there's then no need to use the pager.

I still have stomach ache with the short view, but this may be a personal thing. When I watch myself using it (or the pager) I notice I must read a lot more than in the long view because the eye must move in two directions to find something, vertical and horizontal - in the long view you have only the vertical direction. And when you have a lot of files with very similar names or files with even the same names and similar extensions I need longer to find a file.

An aggravating factor in clifm is that there's no visible cursor you could move from file to file. So the eyes must hold their reading position themselves. That can be difficult on a screen full of text. (This could now even be more important because with these new tight columns you get more text when you have many files.) When you switch the workspace to check something in another directory and come back you must find your reading position again. Sometimes not easy in the short view.

I tend to use the `ft` command more and more in the last time to reduce the amount of text. I have even an F alias meaning `ft .+` (remove a filter) and some more aliases for typical filters.

BTW: I thought also about common wildcards for the `ft` command instead of regexps. I know, a regexp is very exact, you can express almost everything and much more than just wildcards, I'm also fit with regexps, I have no typing problem with them, but wildcards are on the other hand just common for filenames and handy for the very most cases, everyone knows and uses them. I see myself always typing `ft *.` and at the `.` I suddenly notice, damned, it needs a regexp here, not a wildcard.

A second thing with `ft` is that a filter remains active until you remove it. That's sometimes surprising: you set a filter, you do something, you get a telephone call, you come back and then you change the directory by typing `..` or you jump somewhere - when you did this you see probably nothing anymore because everything in the new current directory is filtered. OK, you have then a message line "Showing 0/32 files", so perhaps you can remember, ah, the damned filter is still active. But may also be you can't remember.

This lead me to two ideas:
a) binding a filter on the directory, when you leave it you leave also the filter and when you come back you have the filter again
b) making a filter visible by a prominent mark you can't overlook, for instance a big `[F]` or `[ft]` in the prompt

**Reply by leo-arch** [reply]
So, summing up:

1. Allow the files filter to understand wildcards instead of just REGEXes.
2. Allow the files filter to be active for the current directory only.
3. Add a prompt mark whenever the filter is active.

`1` can be done, of course, but needs some work.
`2` is already in my TODO list. It can be done via autocommands.
`3` should be pretty simple to implement.

I'll let you know when I have something.

Btw, to unset the filter the proper expression is `ft unset`. `ft .+` is still a filter: it allows all files. Now, from a developer's viewpoint, this second expression is expensive, because you still run the filter check for each file.

**Reply by muellerto** [reply]
Just ideas! I was just thinking aloud. And thinking has not finished.

Remember that wildcards have different smells. On Windows `*.*` includes also files without extensions and files with only an extension (dot files). So `*.*` and `*` are indeed synonyms. It's not always consequent. On Linux this is very different, when you write a dot it means a dot. So this leads to the question: Which wildcards? (One advantage is indeed there for regexps: they are always clear.)

`3` would indeed be a help for `ft` power users. (And I would then have noticed that `ft .+` is still an active filter :) )

**Reply by leo-arch** [reply]
Point `3` is done.

**Reply by leo-arch** [reply]
Point `2` is done. Now you can set a files filter per directory via [autocommands](https://github.com/leo-arch/clifm/wiki/Specifics#autocommands). For example, add this line to your config file:

```sh
autocmd ~/Downloads/** ft=!.*\.part$
```

This will exclude all `.part` files from the `~/Downloads` directory (including all subdirectories).

**Reply by muellerto** [reply]
Oh, thanks a lot. Works indeed.

The idea with the autocmd is good because it has already all we need. But you must change the config file for that autocmd.

Another approach could be to have a possibility to construct such an autocmd interactively at the prompt. We need a new command for that because `autocmd` is not a command. Let's say we have a construction like `auto ft regexp`. Then we could use this to _add a new entry into the autocmd-list_ for the current directory which executes an ordinary `ft regexp`. And this entry exists until the user calls `auto ft unset` in this directory or as long as clifm runs. (Note: this could also be useful to extend this concept on sorting or whatever.) Ordinary autocmds in the config file could still be used. A problem is that these interactive autocmds are then volatile and disappear when you reread the config file.

**Reply by leo-arch** [reply]
I see, and it's doable. But there are a few implementation details:

1. When writing an ordinary autocmd entry, you can construct the target directory in three different ways: **a)** match the target exactly (`/DIR`), **2)** match the target's subdirectories, excluding the target itself (`/DIR/*`), and **3)** match both the target and its subdirectories (`/DIR/**`). By omitting the target field from the new `auto` command we loose this flexibility (but we get a more simple syntax in return). Briefly put, I'm not sure about restricting the `auto` command to the current directory.
2. The list of autocmds is read from top to bottom, and the first matching entry is executed, excluding the remaining ones. This means that a new autocmd entry should be not _appended_ to the list, but _prepended_, so that it will take precedence over the remaining entries. This one is rather a development caveat, but it's worth mentioning.
3. Autocommands are reloaded when rereading the config file: true. Indeed, this is the case for most options set from the command line. However, it can be worked around (for example, by keeping a list of options modified from the command line).

**Reply by muellerto** [reply]
1) You could allow an optional directory parameter: `auto [DIR] cmd`. And when the user doesn't provide one it should implicitly support the most typical case: `/DIR/*` (as I understand it).

2) I know these paths things are always complicated, especially with simple C.
**a**) I would allow only one `ft` command in the list for a given directory specification (the same with sorting). When the user types in a second `auto ft` with the same directory specification the existing entry in the list must be changed. This will especially be the case when the user is experimenting with regexps.
**b**) OK, you must then handle such cases where the directory specifications of multiple entries are not equal but also not independent. For instance you have already an entry for a directory including its sub directories and then comes a second entry for a concrete sub directory among these sub directories. - But this case is clear: the more concrete a directory is specified, or the less general it is, the more it is important. `/a/b/c/**` is more general than `/a/b/c/d/e/*`, this means: yes, `/a/b/c/**` affects the whole directory tree under `/a/b/c` but in the special case that you navigate to `/a/b/c/d/e` it has an own entry which comes on top of the general entry.
**c**) You need then still a rule for the special case that you get two entries for `/a/b/c/*` and `/a/b/c/**`. In this case I would define that the first one is more important. This means: all the files and sub directories under `/a/b/c` are generally handled by the second entry but the stuff directly under `/a/b/c` has an own entry.
**d**) It gets complicated when you allow even more wildcards also in the middle of the paths like `/a/b/*/d/e/**`.

This means it would be good to order the entries not by when they have been created but by the length of their paths. A big advantage of such an ordered list is: you need to read it only once from beginning. When you find matches for the current directory you know immediately the order of the matching entries and can execute them from the most general to the most concrete one.

**Reply by leo-arch** [reply]
The first implementation of the `auto` command is ready. Run `auto --help` for details.

**Reply by muellerto** [reply]
Yeah! (You are so fast! .....)

I couldn't check a lot until now (because it's weekend), especially overlapping rules, but my first experiments look really great. This directory based autocmd approach is indeed very powerful, I need still some time to understand all the possibilities but you gave already good examples in the help screen.

**Reply by leo-arch** [reply]
A few things to keep in mind:

1. Options set via the `auto` command are valid **only for the current directory**. Allowing wildcards, or any other way of specifying the target directory, is a hierarchy mess (really prone to misbehaviours). If you need wildcards, use the config file.
2. It would be nice to let the user know when an autocommand has modified the current set of options (maybe a little mark in the prompt or a prompt message, something like `-> Autocmd`).
3. It would also be nice to add a `list` subcommand to list current temporary autocommands, displaying both the target directory and the list of options.
4. The list of temporary autocommands is lost after reloading the config file. Working on it.

**Reply by leo-arch** [reply]
Point `4` solved. But still needs to be checked for stability.

**Reply by leo-arch** [reply]
Point `2` is done. Added a little prompt message every time you change into a directory where an autocommand has been executed. Just a friendly reminder (I hope it's not too intrusive).

**Reply by muellerto** [reply]
Point 2: the prompt message - the idea is good but I have _always_ at least one autocmd in my clifm configuration. It's workspace related. So I have indeed always this message and it looks always equal. Probably it would be good to exclude autocmds from building the message which are not explicitly related to a directory or path.

And, second idea, it would be good to get a hint what the autocmd does. I mean not the entire definition, but a hint like

-> Autocmd [ft, st]

so that the user is informed which aspects of the view are affected.

**Reply by muellerto** [reply]
Point 3 is good. I also thought about this. A list which is filtered for the cwd. So that the user can see what is relevant here on this place.

**Reply by leo-arch** [reply]
Removed the `Autocmd` prompt message until I find a better way to handle these notifications.

Btw, I added validation for both autocommand option names and values.

**Reply by leo-arch** [reply]
The autocommand prompt message includes now a description, which can be customized. Add these lines to your config file:

```sh
# Inform about autocommand options set for the current directory
# Available values:
# false: Do not print any message.
# mini: Minimal message (no options details).
# short: Add a list of modified option names (without values).
# long: Like 'short', but adding values to each option (default).
;InformAutocmd=long
```

**Reply by muellerto** [reply]
Yeah!

Something else: I tried to add a global autocmd line in my config specifying a sort order for some directories. But I failed. I have on my disk several directories containing hundreds of log files and these directories are always named Log, so my general idea is

`autocmd /**/Log/* st=mtime`

This seems to be allowed but it doesn't work. OK, the `**` is probably difficult, so I tried also better local or global paths:

`autocmd ~/work/Log/* st=mtime`
`autocmd /home/muellerto/work/Log/* st=mtime`

But this all doesn't work. When I use the interactive `auto` command it does the job as expected so there must be a difference between `autocmd` and `auto`. (Ah: `auto` knows always it's concrete directory and has no wildcards.)

**Reply by leo-arch** [reply]
`autocmd /**/Log/*` will just not work. The double asterisk is not even an actual glob expression, but an extension which **only works at the end of paths**, meaning `match all subdirectories (recursively) including the specified directory itself`. For example, `~/Log/**` will match `~/Log` and _all_ directories in it.

Both `~/work/Log/*` and `/home/muellerto/work/Log/*` do work, but probably not as you expected: they match all subdirectories in `~/work/Log` (**not** recursively), _excluding_ `~/work/Log` itself. I guess in this case you want `~/work/Log/**` instead.

Take a look at the [documentation](https://github.com/leo-arch/clifm/wiki/Specifics#autocommands).

**Comment by muellerto** [comment]
In my case lines like `autocmd ~/office/* st=mtime` or `autocmd ~/office/** st=mtime` do not sort at all :( even on Linux. And I have a lot of files there. The result looks like this:

```
 1 AXA                                     1 d/rwx.r-x.r-x 2022-11-28 17:39:43m    4096
 2 Adressänderungen                        1 d/rwx.r-x.r-x 2011-07-09 09:59:48m    4096
 3 Anfahrten                               4 d/rwx.r-x.r-x 2019-08-25 11:51:46m    4096
 4 Blog                                    3 d/rwx.r-x.r-x 2016-01-05 17:17:08m    4096
 5 Briefe                                 56 d/rwx.r-x.r-x 2024-05-25 18:44:19m    4096
 6 Coros                                   2 d/rwx.r-x.r-x 2023-03-24 17:03:54m    4096
 7 Covid                                   2 d/rwx.r-x.r-x 2021-11-30 16:09:40m    4096
 8 Texte                                  10 d/rwx.r-x.r-x 2022-08-10 17:40:58m    4096
 9 Finanz                                  3 d/rwx.r-x.r-x 2024-08-02 16:22:30m    4096
10 HP15C                                   5 d/rwx.r-x.r-x 2014-07-06 16:57:06m    4096
11 Krankenkasse                            8 d/rwx.r-x.r-x 2024-04-12 16:17:53m    4096
12 Passbilder                              2 d/rwx.r-x.r-x 2015-05-26 17:21:00m    4096
13 Profil                                  1 d/rwx.r-x.r-x 2015-11-16 17:40:59m    4096
14 SolNet                                  1 d/rwx.r-x.r-x 2019-10-28 17:26:43m    4096
15 Tabellen                               14 d/rwx.r-x.r-x 2024-10-06 17:33:42m    4096
16 Tourenberichte                          6 d/rwx.r-x.r-x 2012-08-26 19:36:22m    4096
17 Amphibia.txt                            - ./rw-.r--.r-- 2013-11-07 16:50:08m    3181
18 Arch-Install-Scripts.odt                - ./rw-.r--.r-- 2024-01-04 16:47:45m   41453
19 FP3-Schrauben.pdf                       - ./rw-.r--.r-- 2020-12-16 14:38:12m   13649
20 Fluege.txt                              - ./rw-.r--.r-- 2018-11-27 17:54:24m    1230
21 H9i-Rechnung.pdf                        - ./rw-.r--.r-- 2021-01-27 17:14:23m  357593
22 Krankenkassenkarte.png                  - ./rw-.r--.--- 2022-06-13 16:56:26m 2781089
23 Lindenhof Zeugnis.png                   - ./rw-.r--.--- 2021-03-16 11:55:51m 3375940
24 Pens.txt                                - ./rw-.r--.r-- 2009-06-04 17:22:08m     173
25 passwords.png                           - ./rw-.r--.r-- 2021-06-24 15:12:20m   12744
26 Suunto_Spartan_Ultra_UserGuide_DE.pdf   - ./rw-.r--.r-- 2018-11-27 17:50:42m 2492152
27 Training.odt                            - ./rw-.r--.r-- 2010-03-08 18:09:24m    9381
28 UE.odt                                  - ./rw-.r--.r-- 2013-12-15 10:52:10m   27311
29 Z30.pdf                                 - ./rw-.r--.r-- 2014-02-08 11:21:46m   24060
───────────────────────────────────────────────────────────────────────────────────────── 
→ Autocmd [lv=1]
T507[1] 17:22 muellerto:thot ~/office
<0> $
```
As I said the long view is a clifmrc defined autocmd based on the workspace. Interesting is that the sorting I defined one line below is not mentioned in the Autocmd message line. Also the short view is not sorted as expected.

And  `autocmd ~/office/* st=mtime rev` produces an error message, it seems not to be allowed.

I use an ordinary build with GIT sources from right now.

**Reply by leo-arch** [reply]
Quick reply:

`autocmd ~/office/* st=mtime rev` won't work, simply because it is not a valid syntax. You want `autocmd ~/office/* st=mtime sr=1`, where `sr` stands for sort-reverse.

I'll take a look at the rest of the stuff.

**Reply by leo-arch** [reply]
Ok. I've modified the main logic of the autocommands function: all matching entries are executed, instead of just the first one.

This should fix your issue.

Btw, the prompt message for autocommands is disabled until I find a way to properly display options for multiple autocmd entries. The main thing is this: should we print a) **separate messages** for each matching autocmd (more intrusive, but clearer), or rather b) a **single message summarizing the options set by all matching autocmds** (less intrusive, less clear)? Another advantage of the first approach is that it would allow us to specify whether a given entry is temporal (i.e. set via the `auto` command) or permanent, which could be useful.

Forgot to mention: autocmds are evaluated from top to bottom, and this means that the more general a pattern or rule is, the more on top of the list should it be placed. For example, this:

```sh
autocmd ~/Documents lv=0
autocmd @ws1 lv=1
```
won't work, because the first rule is overridden by the second one, since both of them match the current directory. Instead, you want this:

```sh
autocmd @ws1 lv=1
autocmd ~/Documents lv=0
```

**Reply by muellerto** [reply]
Now I got it! I had a misunderstanding in the meaning of `dir`, `dir/*` and `dir/**` in the directory specification of an `autocmd`. I always had a wrong `*` at the end, as seen in my ~/office example above with the listing.

OK, with that I can now lay sorting instructions on concrete paths pointing on my several Log directories. I have seven at the moment, dynamically changing from week to week with the number of branch specific working copies I have. But I can't make a general rule, once for all, also for future ones.

Ah, update: a single * in the middle of a path works, `~/work/*/test_cfg/Log`, so I need indeed only one `autocmd` line for all my working copies.

**Reply by leo-arch** [reply]
> Ah, update: a single * in the middle of a path works, ~/work/*/test_cfg/Log, so I need indeed only one autocmd line for all my working copies.

Good to know. However, there are still a few regressions (due to the new implementation) I need to work out. Please report whatever comes out.

**Comment by leo-arch** [comment]
Replaced **glob**(3) by **fnmatch**(3) when checking autocmds: it's faster and more flexible. Now you can match your `Log` directory more generally as follows:

```sh
autocmd */Log ...
```
instead of

```sh
autocmd ~/work/*/test_cfg/Log ...
```
which is more specific.

**Reply by muellerto** [reply]
That's great news. This extends the usability a lot and makes it much more intuitive. Will test this on Monday.

**Reply by leo-arch** [reply]
Added a new option (`full`) to `InformAutocmd`: it's like `long`, but multiple matching autocommands are printed on separate lines (also, a `T` is appended to mark temporary autocommands, i.e., set via the `auto` command).

**Reply by leo-arch** [reply]
Added `list` subcommand to the `auto` command to list available autocommands.

**Reply by leo-arch** [reply]
Added a new option (`prompt`) to `InformAutocmd`: it prints a little prompt indicator (a gray `A`) at the left of the prompt. The color of this indicator is controlled by the new `ac` color code in the color scheme file. This is the new default value for `InformAutocmd`.

I think we have more balance in this way: the indicator is barely noticeable, and with the `auto list` command you can get all the information you need if required.

**Reply by leo-arch** [reply]
Currently available values for `InformAutocmd`: `none`, `mini`, `short`, `long`, `full`, and `prompt` (replaced `false` by `none`).

**Reply by muellerto** [reply]
Currently I use full. OK, long would show the same.

Generally I would say the prompt would be the right place. But as I said before, I have always a workspace related `autocmd` (coloring) because I tend to use always the same workspace for the same tasks and it's very handy to recognize easily where you are. So the A is indeed always there.

**Reply by leo-arch** [reply]
> OK, long would show the same.

This is true in the case of a **single** matching autocmd. But when **multiple** autocmds are matched, both options behave differently: while `long` merges options coming from all matching autocmds into a single line, `full` prints separate lines for each autocmd (adding an ending `T` to the line if the autocmd is temporal, i.e., set via the `auto` command).

> So the A is indeed always there.

The `prompt` option might not be useful in your specific use case, because you always have autocmds set, but it's pretty sleek when you use autocmds only for specific directories. Be it as it may, this is why `InformAutocmd` supports 7 different behaviors: to cover different use cases.

**Reply by muellerto** [reply]
I have an `autocmd` in clifmrc (the coloring) and another entry made with the `auto` command (the filter). Now I have the following messages:

```
→ Showing 4/238 files
→ Active filter: .+\.exe
→ Autocmd [cs=solarized]
→ Autocmd [ft]T
```

questions:
- why is the `cs` option full explicit but the `ft` option isn't?
- what's the `T` after `[ft]`? I checked this, it's always there and it's always a `T`.

As I said, I have `InformAutocmd=full`.

**Reply by leo-arch** [reply]
The `ft` option isn't expanded to its value because the filter is already displayed by the `Active filter` message (no need to print the same thing twice).

The `T` after the autocmd options means that this autocmd is temporal, that is, set via the `auto` command.

**Reply by muellerto** [reply]
One good trait of this application is that everything - how strange it may be - is still explainable :)

**Reply by leo-arch** [reply]
Thanks. However, an even better application does not need explanations, because it is self-explaining. Sometimes, the need for explanation reflects a bad design decision. I'll keep working on this.

**Reply by leo-arch** [reply]
The current entry for `InformAutocmd` is this:

```sh
...
# Run autocommand files (.cfm.in and .cfm.out) found in the current
# directory. For more information consult the manpage.
;ReadAutocmdFiles=false

# Inform about autocommand options set for the current directory
# Available values:
# 'none': Do not print any message.
# 'mini': Minimal message (no options details).
# 'short': Add the list of autocommand options (without values).
# 'long': Like 'short', but displays the value of each option (multiple
#   matching autocommands are merged into a single line).
# 'full': Like 'long', but multiple autocommands are printed on separate
#   lines. Also, a 'T' is appended to mark temporary autocommands (i.e.,
#   set via the 'auto' command).
# 'prompt': Print a little notification (a gray 'A') in the prompt.
#   The color code used for this notification is 'ac' (see the color scheme file).
#   If Notifications is set to false for the current prompt (see the prompts file),
#   nothing is printed (you can use the '\y' prompt code in the prompts file to
#   manually set this notification).
;InformAutocmd=prompt

# Run in read-only mode (internal commands able to modify the file
# system are disabled). Consult the manpage for the list of affected
# commands.
;Readonly=false
...
```

**Reply by muellerto** [reply]
I made a spelling error and wrote `auto lit` - this says
```
clifm: xnrealloc failed to allocate 0x104 bytes
free(): double free detected in tcache 2
```
and gives then a segmentation fault :(

**Reply by leo-arch** [reply]
Great. I'll fix this right now.

**Reply by leo-arch** [reply]
Can you reproduce it? Haven't seen any realloc error.

EDIT: Forget about it. Reproduced it.

**Reply by leo-arch** [reply]
Should be fixed now.

**Reply by muellerto** [reply]
On my Windows machine it didn't happen this morning. Tried it several times, was always good. I have the original version from yesterday still on Linux.

**Reply by muellerto** [reply]
So, I had it again yesterday on my Linux machine with now two days old binaries and it seemed to be reproducible. Now with actual binaries the problem is indeed gone :)

**Reply by leo-arch** [reply]
We can consider this issue solved then. Thanks @muellerto.
